File: bandage/src/main/java/com/panda912/bandage/utils/ActivityManager.java
package com.panda912.bandage.utils;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.lang.ref.SoftReference;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Created by panda on 2020/12/10 15:29
 */
public class ActivityManager implements Application.ActivityLifecycleCallbacks {
  private static volatile ActivityManager sInstance;

  private final List<SoftReference<Activity>> mActivityList = new CopyOnWriteArrayList<>();

  private ActivityManager() {
  }

  public static ActivityManager getInstance() {
    if (sInstance == null) {
      synchronized (ActivityManager.class) {
        if (sInstance == null) {
          sInstance = new ActivityManager();
        }
      }
    }
    return sInstance;
  }

  @Nullable
  public Activity getCurActivity() {
    if (mActivityList.isEmpty()) {
      return null;
    }

    SoftReference<Activity> softReference = mActivityList.get(mActivityList.size() - 1);
    if (softReference == null) {
      return null;
    }
    return softReference.get();
  }

  public boolean isDestroyed(Activity activity) {
    return activity == null || activity.isFinishing() || activity.isDestroyed();
  }

  @Override
  public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle savedInstanceState) {
    mActivityList.add(new SoftReference<>(activity));
  }

  @Override
  public void onActivityStarted(@NonNull Activity activity) {

  }

  @Override
  public void onActivityResumed(@NonNull Activity activity) {

  }

  @Override
  public void onActivityPaused(@NonNull Activity activity) {

  }

  @Override
  public void onActivityStopped(@NonNull Activity activity) {

  }

  @Override
  public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle bundle) {

  }

  @Override
  public void onActivityDestroyed(@NonNull Activity activity) {
    for (SoftReference<Activity> softReference : mActivityList) {
      if (softReference.get() == activity) {
        mActivityList.remove(softReference);
        return;
      }
    }
  }
}

File: bandage/src/main/java/com/panda912/bandage/utils/ActivityManager.kt
package com.panda912.bandage.utils

import android.app.Activity
import android.app.Application
import android.os.Bundle
import java.lang.ref.SoftReference
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Created by panda on 2021/12/16 10:13
 */
class ActivityManager : Application.ActivityLifecycleCallbacks {

  private val activityList = CopyOnWriteArrayList<SoftReference<Activity>>()

  fun getCurActivity(): Activity? {
    if (activityList.isEmpty()) {
      return null
    }
    return activityList[activityList.size - 1]?.get()
  }

  fun isDestroyed(activity: Activity?): Boolean =
    activity == null || activity.isFinishing || activity.isDestroyed

  override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {
    activityList.add(SoftReference(activity))
  }

  override fun onActivityStarted(activity: Activity) {

  }

  override fun onActivityResumed(activity: Activity) {

  }

  override fun onActivityPaused(activity: Activity) {

  }

  override fun onActivityStopped(activity: Activity) {

  }

  override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {

  }

  override fun onActivityDestroyed(activity: Activity) {
    for (softReference in activityList) {
      if (softReference.get() === activity) {
        activityList.remove(softReference)
        return
      }
    }
  }

  companion object {
    @Volatile
    private var instance: ActivityManager? = null

    fun getInstance(): ActivityManager =
      instance ?: synchronized(this) {
        instance ?: ActivityManager().also { instance = it }
      }

  }
}